### PR TITLE
Add support Provider Gateway topology intentions on VCD 10.5.1 (NAT, FW, route advertisement)

### LINKED ISSUE #1208: configure NAT and Firewall Service Intention in vcd_external_network_v2 (Provider Gateway)
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

It is possible to configure "NAT and Firewall Service Intention" in VCD v10.5.1+ UI.

### New or Affected Resource(s)

* vcd_external_network_v2

### Terraform Configuration (if it applies)

Proposal of a new argument

```hcl
service_intention = "PROVIDER_AND_EDGE_GATEWAY" // Optional (VCD 10.5.1+) / This setting controls where Firewall and NAT will be visible for configuration. By default, Firewall and NAT will only be configurable on Edge Gateways. Default: "EDGE_GATEWAY"
```

### References

grafik

### LINKED ISSUE #1238: vcd_external_network_v2 Route Advertisement Strict, Flexible and All
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The new version of VCD (v10.5.1) now offers to setup "Advertisement Strict", "Advertisement Flexible" and "All Networks Advertised". Currently it defaults to "Advertisement Strict", which we might need to change in the future.

### New or Affected Resource(s)

* vcd_external_network_v2

### Terraform Configuration (if it applies)

Proposal of a new argument

```hcl
route_advertisement = "STRICT" // Optional (VCD 10.5.1+) / This setting controls which networks will be advertised. Default: "STRICT"
```

### PR DESCRIPTION
**Pull request by Didainius**
This PR adds Topology intention configuration for Firewall, Nat and Route advertisement.

Ref: https://github.com/vmware/terraform-provider-vcd/issues/1105#issuecomment-1933684183
Closes #1238
Closes #1208

### CODE REVIEW

File: website/docs/r/external_network_v2.html.markdown [hunk 1]
```diff
@@ -195,6 +217,26 @@ The following arguments are supported:
...
 * `vsphere_network` - (Optional) One or more blocks of [vSphere Network](#vspherenetwork)..
 * `nsxt_network` - (Optional) NSX-T network definition. See [NSX-T Network](#nsxtnetwork) below for details.
+* `nat_and_firewall_service_intention` - (Optional; *TBC+*; *VCD 10.5.1+*) Configure intentions for
+ NAT and Firewall rule configuration:
+ * `EDGE_GATEWAY` - Allow management of NAT and firewall rules only on Edge Gateways. This is the
+   default behavior.
+ * `PROVIDER_GATEWAY` - Allow management of NAT and firewall rules only on Provider Gateways.
+ * `PROVIDER_AND_EDGE_GATEWAY` - Allow management of NAT and firewall rules on both the Provider and
+ Edge gateways.
+* `route_advertisement_intention` - (Optional; *TBC+*; *VCD 10.5.1+*) Configure intentions for
```
[the remark this answers is not in the document]
Comment (Didainius) [reply website/docs/r/external_network_v2.html.markdown]: Indeed. Fixed.

**Review by lvirbalas — APPROVED** [review]

File: go.mod [hunk 2]
```diff
@@ -66,3 +66,5 @@ require (
 	google.golang.org/grpc v1.60.0 // indirect
 	google.golang.org/protobuf v1.31.0 // indirect
 )
+
+replace github.com/vmware/go-vcloud-director/v2 => github.com/Didainius/go-vcloud-director/v2 v2.17.0-alpha.2.0.20240426055454-19d6098e3f6b
```
Comment (dataclouder) [inline go.mod:70]: reminder to remove this replacement before merging

File: vcd/resource_vcd_external_network_v2.go [hunk 3]
```diff
@@ -154,6 +154,18 @@ func resourceVcdExternalNetworkV2() *schema.Resource {
...
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: "Defines intentions to configure NAT and Firewall rules (only with IP Spaces, VCD 10.5.1+) One of `PROVIDER_GATEWAY`,`EDGE_GATEWAY`,`PROVIDER_AND_EDGE_GATEWAY`",
+			},
+			"route_advertisement_intention": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: "Defines intentions to configure route advertisement (only with IP Spaces, VCD 10.5.1+) One of `IP_SPACE_UPLINKS_ADVERTISED_STRICT`,`IP_SPACE_UPLINKS_ADVERTISED_FLEXIBLE`,`ALL_NETWORKS_ADVERTISED`",
```
Comment (dataclouder) [inline vcd/resource_vcd_external_network_v2.go:167]: Shouldn't we have a validation function here?
Otherwise, we end up with a BAD_REQUEST error

Comment (Didainius) [reply vcd/resource_vcd_external_network_v2.go]: I added validations to both fields `nat_and_firewall_service_intention` and `route_advertisement_intention`. 

Test run worked:
```
=== RUN   TestAccVcdExternalNetworkV2Datasource
--- PASS: TestAccVcdExternalNetworkV2Datasource (16.87s)
=== RUN   TestAccVcdExternalNetworkV2NsxtVrf
--- PASS: TestAccVcdExternalNetworkV2NsxtVrf (81.89s)
=== RUN   TestAccVcdExternalNetworkV2NsxtT0Router
--- PASS: TestAccVcdExternalNetworkV2NsxtT0Router (79.04s)
=== RUN   TestAccVcdExternalNetworkV2Nsxv
--- PASS: TestAccVcdExternalNetworkV2Nsxv (44.48s)
=== RUN   TestAccVcdExternalNetworkV2NsxtSegment
--- PASS: TestAccVcdExternalNetworkV2NsxtSegment (31.78s)
=== RUN   TestAccVcdExternalNetworkV2NsxtConfigError
--- PASS: TestAccVcdExternalNetworkV2NsxtConfigError (1.67s)
=== RUN   TestAccVcdExternalNetworkV2NsxtSegmentIntegration
--- PASS: TestAccVcdExternalNetworkV2NsxtSegmentIntegration (102.35s)
=== RUN   TestAccVcdExternalNetworkV2NsxtIpSpace
--- PASS: TestAccVcdExternalNetworkV2NsxtIpSpace (97.09s)
=== RUN   TestAccVcdExternalNetworkV2NsxtTopologyIntentionEdgeAndStrict
--- PASS: TestAccVcdExternalNetworkV2NsxtTopologyIntentionEdgeAndStrict (35.87s)
=== RUN   TestAccVcdExternalNetworkV2NsxtTopologyIntentionProviderAndFlexible
--- PASS: TestAccVcdExternalNetworkV2NsxtTopologyIntentionProviderAndFlexible (37.07s)
=== RUN   TestAccVcdExternalNetworkV2NsxtTopologyIntentionProviderAndEdgeAllNetworks
--- PASS: TestAccVcdExternalNetworkV2NsxtTopologyIntentionProviderAndEdgeAllNetworks (36.98s)
=== RUN   TestAccVcdExternalNetworkV2NsxtTopologyIntentionIntegration
--- PASS: TestAccVcdExternalNetworkV2NsxtTopologyIntentionIntegration (157.57s)
```

**Review by dataclouder — APPROVED** [review]

**Review by adambarreiro — COMMENTED** [review]
One doubt

File: vcd/resource_vcd_external_network_v2.go [hunk 4]
```diff
@@ -332,6 +348,10 @@ func getExternalNetworkV2Type(vcdClient *VCDClient, d *schema.ResourceData, know
 		return nil, fmt.Errorf("'dedicated_org_id' can only be set when 'use_ip_spaces' is enabled")
 	}
 
+	if vcdClient.Client.APIVCDMaxVersionIs("< 38.1") {
```
[the remark this answers is not in the document]
Comment (Didainius) [reply vcd/resource_vcd_external_network_v2.go:351]: Thanks for noticing. Indeed this was a broken validation. Updated. Checked on multiple versions now and ran external network tests. Looks to be good on older and newer.

**Review by adambarreiro — APPROVED** [review]

File: go.mod (hunk 2 above)
Comment (Didainius) [reply go.mod]: Done.

**Pull request merged**